### PR TITLE
Bootstrap a new history_uuid when force allocating a stale primary

### LINKED ISSUE #26712: Force a new history uuid when force allocating a stale primary
This is a follow up to #26694 which didn't cover that edge case. The rationale is similar - it's unsafe to do ops based recoveries from those primaries

### PR DESCRIPTION
**Pull request by dnhatn**
This commit ensures that we bootstrap a new history_uuid when force
allocating a stale primary. A stale primary should never be the source
of an operation-based recovery to another shard which exists before the
forced-allocation.

Closes #26712

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java [hunk 1]
```diff
@@ -107,25 +112,68 @@ public int hashCode() {
     }
 
     /**
-     * recovery from an existing on-disk store or a fresh copy
+     * Recovery from a refresh copy
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java]: refresh -> fresh?

File: server/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 2]
```diff
@@ -213,6 +221,10 @@ public void testForceStaleReplicaToBePromotedToPrimary() throws Exception {
         ClusterState state = client().admin().cluster().prepareState().get().getState();
         assertEquals(Collections.singleton(state.routingTable().index(idxName).shard(0).primary.allocationId().getId()),
             state.metaData().index(idxName).inSyncAllocationIds(0));
+
+        Set<String> newHistoryUUIds = Arrays.stream(client().admin().indices().prepareStats("test").clear().get().getShards())
+            .map(shard -> shard.getCommitStats().getUserData().get(Engine.HISTORY_UUID_KEY)).collect(Collectors.toSet());
+        assertThat(newHistoryUUIds, everyItem(not(isIn(historyUUIDs))));
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: mabye assert that all new history uuids are the same, just for fun?

**Review by bleskes — APPROVED** [review]
LGTM. Nice that we can do this now via store with little hassle.

**Comment by dnhatn** [comment]
Thanks @bleskes for reviewing.

**Comment by ywelsch** [comment]
Note that if the node crashes between shard initialization and shard recovery, the allocation id (both in the in-sync set as well as on the node's local shard copy) will be adjusted but the history UUID will not, which is unsafe to do. I would have preferred to put the history UUID into the index metadata (on a per-shard level, similar to in-sync-allocation ids), and then realigned the shards history uuid with the index metadata one, avoiding the above problem.

**Comment by bleskes** [comment]
@ywelsch that's a good insight. If we put the history uuid in the index meta data, I think we need to work out the exact semantics of it and how it correlates to the lucene index version (or may we want to move to store it in the state file together with the history uuid). I think we can also consider another alternative - only add the allocation id of the stale copy to the insync set once it has been recovered (potentially clearing the in sync set when the command is run). I also wonder if we should do the same for recovery from snapshot (if I read the code correctly, I *think* we have the same problem)

**Comment by bleskes** [comment]
>  I would also like for the allocate stale primary command to immediately take effect (in terms of resetting the in-sync set).

Agreed.

> We could just make it a singleton set with a random allocation id

I first thought this was a hack, but then I thought that maybe we should set it to `[ "_forced_allocation" ]` which will increase the transparency of what happened until we get back to "normal" (i.e., the primary was started)

**Comment by ywelsch** [comment]
> maybe we should set it to [ "_forced_allocation" 

works for me.

**Pull request merged**